### ISSUE TITLE
Strange collapse issue with the dino loss

### ISSUE THREAD

**Issue by swarajnanda2021** [facebookresearch/dinov2#407]
I've got a strange collapse issue here when I only use the dino loss term alone. I've got a custom implementation of the dino, but I've checked nearly every line possible to see if there is any difference, against the original dino method. I get a collapse immediately, or later, depending on the learning rate.

Strangely it collapses always at the same number: 8.317765235900879 (this changes with the head dimension size).

Here is what this looks like:
![image](https://github.com/facebookresearch/dino/assets/85413978/478f987f-21ce-46b2-be50-172e6dbe02a2)

Secondly, here is what my code looks like:

the loss function is a direct copy and paste from dino's repository, while the method is written in a pytorch lightning module, which is itself an almost copy and paste from another repository.

```ruby
def cosine_scheduler(base_value, final_value, epochs, niter_per_ep,
                     warmup_epochs=0, start_warmup_value=1e-6):
    warmup_schedule = np.array([])
    warmup_iters = warmup_epochs * niter_per_ep
    if warmup_epochs > 0:
        warmup_schedule = np.linspace(start_warmup_value, base_value, warmup_iters)

    iters = np.arange(epochs * niter_per_ep - warmup_iters)
    schedule = final_value + 0.5 * (base_value - final_value) * \
                             (1 + np.cos(np.pi * iters / len(iters)))

    schedule = np.concatenate((warmup_schedule, schedule))
    assert len(schedule) == epochs * niter_per_ep
    return schedule


class DiNO(pl.LightningModule):
    def __init__(
            self,
            feature_size, 
            dinoheadout,
            encoder, 
            temperature_teacher, 
            final_teacher_temp,
            ncrops, 
            momentum_teacher,
            batch_size,
            num_gpus,
            lr,
            initial_weight_decay,
            final_weight_decay,
            warmup_epochs,
            max_epochs,
            iters_per_epoch,
    ):
        super().__init__()
        self.save_hyperparameters(ignore=['encoder'])
        self.automatic_optimization = False
        student_backbone = encoder
        self.teacher_backbone = encoder
        # set teacher_backbone dropoouts to zero
        self.teacher_backbone.attention_dropout, self.teacher_backbone.projection_dropout = 0.0, 0.0

        student_head = DiNOProjection(
                                    in_dim=feature_size,
                                    out_dim=dinoheadout,
                                    use_bn=False,
                                    norm_last_layer=False
                                    )
        teacher_head = DiNOProjection(
                                    in_dim=feature_size,
                                    out_dim=dinoheadout,
                                    use_bn=False,
                                    norm_last_layer=True
                                    )
        
        self.student = MultiCropWrapper(student_backbone, student_head)
        self.teacher = MultiCropWrapper(self.teacher_backbone, teacher_head)
        # teacher and student start with the same weights
        self.teacher.load_state_dict(self.student.state_dict())

        for p in self.teacher.parameters(): p.requires_grad = False


        self.loss = DiNOLoss(
                        out_dim     = dinoheadout,
                        ncrops      = ncrops + 2,
                        warmup_teacher_temp = temperature_teacher,
                        final_teacher_temp  = final_teacher_temp,
                        warmup_teacher_temp_epochs = 30,
                        nepochs     = max_epochs,
                        )

    def configure_optimizers(self):
        # First, ensure that weight decay does not apply to bias terms or norm params
        regularized, not_regularized = [], []
        for n, p in self.student.named_parameters():
            if not p.requires_grad:
                continue
            # we do not regularize biases nor Norm parameters
            if n.endswith(".bias") or len(p.shape) == 1:
                not_regularized.append(p)
            else:
                regularized.append(p)
        param_groups = [{'params': regularized                          },
                        {'params': not_regularized, 'weight_decay': 0.  }]

        lr = self.hparams.lr * (self.hparams.batch_size*self.hparams.num_gpus/256) # add accum gradient here later
        optimizer = LAMB(param_groups, lr=lr)
        
        return optimizer
    
    def setup(self, stage = None):

        # define schedulers based on number of iterations
        niter_per_ep = self.hparams.iters_per_epoch
        self.lr_sch = cosine_scheduler(self.hparams.lr, 
                                       1e-6, 
                                       self.hparams.max_epochs, 
                                       niter_per_ep//self.hparams.num_gpus,
                                       self.hparams.warmup_epochs)
        # weight decay scheduler
        self.wd_sch = cosine_scheduler(self.hparams.initial_weight_decay, 
                                       self.hparams.final_weight_decay,
                                       self.hparams.max_epochs, 
                                       niter_per_ep//self.hparams.num_gpus)
        # momentum scheduler
        self.mm_sch = cosine_scheduler(self.hparams.momentum_teacher, 
                                       1.0,
                                       self.hparams.max_epochs, 
                                       niter_per_ep//self.hparams.num_gpus)

    def training_step(self, batch, batch_idx):
        """
        batch: a list of "2+local_crops_number" tensors
               each tensor is of shape (B, 3, h, w)
        """

        imgs, _ = batch

        opt = self.optimizers()
        # update learning rate, weight decay
        for i, param_group in enumerate(opt.param_groups):
            param_group['lr'] = self.lr_sch[self.global_step]
            if i == 0: # only the first group is regularized
                param_group['weight_decay'] = self.wd_sch[self.global_step]

        teacher_output = self.teacher(imgs[:2])
        student_output = self.student(imgs)
        loss = self.loss(student_output, teacher_output, self.current_epoch)

        opt.zero_grad()
        self.manual_backward(loss)

        # Perform gradient clipping and freeze last layer of student if first epoch
        torch.nn.utils.clip_grad_norm_(self.student.parameters(), 0.5)
        if self.current_epoch < 1:
            for n, p in self.student.named_parameters():
                if 'last_layer' in n:
                    p.grad = None
        opt.step()

        # EMA update for the teacher
        m = self.mm_sch[self.global_step]
        for ps, pt in zip(self.student.parameters(), self.teacher.parameters()):
            pt.data.mul_(m).add_((1-m)*ps.detach().data)

        # Logging
        self.log('loss', loss, prog_bar=True)
        self.log('lr',self.trainer.optimizers[0].param_groups[0]['lr'],prog_bar=True)

```

The hyperparameters I have used are as follows:

```ruby
batch_size: 24
dinoheadout: 4096
feature_size: 384
final_teacher_temp: 0.07
final_weight_decay: 0.4
initial_weight_decay: 0.04
iters_per_epoch: 55416
lr: 0.0005
max_epochs: 300
momentum_teacher: 0.996
ncrops: 4
num_gpus: 4
temperature_teacher: 0.001 (Have been trying to reduce this from 0.02 to 0.01 to 0.005 and now 0.001. No help)
warmup_epochs: 10

```

My dataset contains 1.2 million images, and I am using a transformer backbone.

**Comment by MohammedSB** [comment]
In this part
```
student_backbone = encoder
self.teacher_backbone = encoder
```
you are setting the student and teacher backbone to the same model. You are assigning by reference here, meaning if you change the student model's weight the teacher will also change. 
You should create another new, independent model for the teacher instead.

**Comment by swarajnanda2021** [comment]
I see. Let me give this a try.

**Comment by swarajnanda2021** [comment]
The matter is resolved. I had issues with both my vision transformer and an over-reliance on the lightning library. I have had success sticking to pytorch raw.

**Comment by cgebbe** [comment]
@swarajnanda2021 : Is the implementation using pytorch lightning publicly available?

**Comment by MyHogsmeade** [comment]
> The matter is resolved. I had issues with both my vision transformer and an over-reliance on the lightning library. I have had success sticking to pytorch raw. The deepcopy versus '=' also solved a massive bug too.

hi, so, what does a normal final loss curve look like?

**Comment by natwille1** [comment]
same question as above!

**Comment by swarajnanda2021** [comment]
I don't know. It depends on the center momentum start value, the size of the ViT, and whether you have set the grad norm last layer to true or false. The loss curve varies for these parameters only in the start, then there is a monotonic behavior.

**Comment by Sann5** [comment]
> Strangely it collapses always at the same number: 8.317765235900879 (this changes with the head dimension size).

This is approximately 2*ln(K), where K is the head prototype dimension. The factor of 2 is because the dino loss is computed for the global and the local crops. Essentially, this means that the output distribution over the prototypes is uniform, i.e., you were experiencing collapse to the uniform distribution.
